File: ios/Songitude/Songitude/GeoUtils.swift
import Foundation
import CoreLocation

/// Geometry helpers shared by the audio engine (containment) and the map overlay.
enum GeoUtils {

    static func distance(_ a: CLLocationCoordinate2D, _ b: CLLocationCoordinate2D) -> CLLocationDistance {
        CLLocation(latitude: a.latitude, longitude: a.longitude)
            .distance(from: CLLocation(latitude: b.latitude, longitude: b.longitude))
    }

    /// Even-odd ray casting on lat/lng. Matches the editor's `pointInPolygon`.
    static func pointInPolygon(_ p: CLLocationCoordinate2D, ring: [CLLocationCoordinate2D]) -> Bool {
        guard ring.count >= 3 else { return false }
        var inside = false
        var j = ring.count - 1
        for i in 0..<ring.count {
            let yi = ring[i].latitude, xi = ring[i].longitude
            let yj = ring[j].latitude, xj = ring[j].longitude
            let intersect = (yi > p.latitude) != (yj > p.latitude) &&
                p.longitude < (xj - xi) * (p.latitude - yi) / (yj - yi + 1e-15) + xi
            if intersect { inside.toggle() }
            j = i
        }
        return inside
    }

    /// Is `coord` inside `shape`, after applying a display/test `offset` to the shape's geometry?
    static func contains(_ shape: SoundShape, coord: CLLocationCoordinate2D, offset: CoordinateOffset) -> Bool {
        switch shape.type {
        case .circle:
            guard let c = shape.centerCoord, let r = shape.radius else { return false }
            return distance(offset.apply(c), coord) <= r
        case .polygon:
            let ring = shape.ringCoords.map { offset.apply($0) }
            return pointInPolygon(coord, ring: ring)
        }
    }
}

/// A lat/lng shift used by the "re-center map over me" debug feature, so a map authored for
/// one city can be tested wherever the user physically is.
/// A rigid move-and-turn of a whole walk: the authored anchor is carried onto the listener and
/// spun so the direction the anchor faced now points where the listener is facing. Distances and
/// angles between shapes are preserved, so the walk plays exactly as composed — just somewhere else.
struct WalkTransposition {
    let from: CLLocationCoordinate2D      // authored anchor
    let to: CLLocationCoordinate2D        // where the listener stands
    let turn: Double                      // radians clockwise, listener heading − anchor heading

    init(anchor: WalkAnchor, listener: CLLocationCoordinate2D, heading: Double) {
        from = anchor.coord
        to = listener
        turn = (heading - anchor.heading) * .pi / 180
    }

    private static let metresPerDegreeLat = 110_540.0
    private static func metresPerDegreeLng(at lat: Double) -> Double {
        111_320.0 * cos(lat * .pi / 180)
    }

    func apply(_ c: CLLocationCoordinate2D) -> CLLocationCoordinate2D {
        // Local east/north offset from the anchor, in metres.
        let east = (c.longitude - from.longitude) * Self.metresPerDegreeLng(at: from.latitude)
        let north = (c.latitude - from.latitude) * Self.metresPerDegreeLat
        // Turn the bearing of that offset by `turn` (clockwise from north).
        let cs = cos(turn), sn = sin(turn)
        let east2 = east * cs + north * sn
        let north2 = north * cs - east * sn
        return CLLocationCoordinate2D(
            latitude: to.latitude + north2 / Self.metresPerDegreeLat,
            longitude: to.longitude + east2 / Self.metresPerDegreeLng(at: to.latitude))
    }

    func apply(_ pair: [Double]) -> [Double] {
        guard pair.count == 2 else { return pair }
        let c = apply(CLLocationCoordinate2D(latitude: pair[0], longitude: pair[1]))
        return [c.latitude, c.longitude]
    }
}

extension SoundMap {
    /// A copy of this walk moved and turned onto the listener. Radii are untouched — only positions
    /// move — so every area keeps its authored size and relative bearing.
    func transposed(_ t: WalkTransposition) -> SoundMap {
        var copy = self
        if let c = center { copy.center = t.apply(c) }
        copy.shapes = shapes.map { shape in
            var s = shape
            if let c = s.center { s.center = t.apply(c) }
            if let pts = s.points { s.points = pts.map(t.apply) }
            return s
        }
        // Routes travel with the walk too — a suggested path left behind where the walk was
        // authored would point the listener at nothing.
        copy.routes = routes?.map { route in
            var r = route
            r.points = r.points.map(t.apply)
            return r
        }
        return copy
    }
}

struct CoordinateOffset: Equatable {
    var dLat: Double = 0
    var dLng: Double = 0

    static let none = CoordinateOffset()

    func apply(_ c: CLLocationCoordinate2D) -> CLLocationCoordinate2D {
        CLLocationCoordinate2D(latitude: c.latitude + dLat, longitude: c.longitude + dLng)
    }

    /// Offset that moves `mapCenter` on top of `userLocation`.
    static func recentering(mapCenter: CLLocationCoordinate2D, onto userLocation: CLLocationCoordinate2D) -> CoordinateOffset {
        CoordinateOffset(dLat: userLocation.latitude - mapCenter.latitude,
                         dLng: userLocation.longitude - mapCenter.longitude)
    }
}

File: android/app/src/main/java/com/brianellissound/songitude/model/GeoUtils.kt
package com.brianellissound.songitude.model

import kotlin.math.abs
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt

/**
 * Geometry shared by the audio engine (containment) and the map overlay.
 * Ported from ios/Songitude/Songitude/GeoUtils.swift — the containment test in particular must
 * agree with the editor's `pointInPolygon`, or a walk sounds different here than it was composed.
 */
object GeoUtils {

    private const val EARTH_RADIUS_M = 6_371_008.8

    /** Great-circle distance in metres, matching CLLocation.distance(from:) closely enough that a
     *  radius test lands on the same side of the boundary. */
    fun distance(a: LatLngD, b: LatLngD): Double {
        val lat1 = Math.toRadians(a.lat)
        val lat2 = Math.toRadians(b.lat)
        val dLat = lat2 - lat1
        val dLng = Math.toRadians(b.lng - a.lng)
        val h = sin(dLat / 2) * sin(dLat / 2) +
                cos(lat1) * cos(lat2) * sin(dLng / 2) * sin(dLng / 2)
        return 2 * EARTH_RADIUS_M * atan2(sqrt(h), sqrt(1 - h))
    }

    /** Even-odd ray casting on lat/lng. Matches the editor's `pointInPolygon` exactly, including
     *  the 1e-15 guard that keeps a horizontal edge from dividing by zero. */
    fun pointInPolygon(p: LatLngD, ring: List<LatLngD>): Boolean {
        if (ring.size < 3) return false
        var inside = false
        var j = ring.size - 1
        for (i in ring.indices) {
            val yi = ring[i].lat; val xi = ring[i].lng
            val yj = ring[j].lat; val xj = ring[j].lng
            val intersect = (yi > p.lat) != (yj > p.lat) &&
                    p.lng < (xj - xi) * (p.lat - yi) / (yj - yi + 1e-15) + xi
            if (intersect) inside = !inside
            j = i
        }
        return inside
    }

    /** Is [coord] inside [shape], after applying a display/test [offset] to the shape's geometry? */
    fun contains(shape: SoundShape, coord: LatLngD, offset: CoordinateOffset): Boolean =
        when (shape.type) {
            ShapeType.CIRCLE -> {
                val c = shape.centerCoord
                val r = shape.radius
                if (c == null || r == null) false else distance(offset.apply(c), coord) <= r
            }
            ShapeType.POLYGON -> pointInPolygon(coord, shape.ringCoords.map { offset.apply(it) })
        }

    /** How far into a circle the listener is, 0 at the centre and 1 at the edge. Used for falloff. */
    fun normalizedRadius(shape: SoundShape, coord: LatLngD, offset: CoordinateOffset): Double {
        val c = shape.centerCoord ?: return 1.0
        val r = shape.radius ?: return 1.0
        if (r <= 0) return 1.0
        return (distance(offset.apply(c), coord) / r).coerceIn(0.0, 1.0)
    }
}

/**
 * A rigid move-and-turn of a whole walk: the authored anchor is carried onto the listener and spun
 * so the direction the anchor faced now points where the listener is facing. Distances and angles
 * between shapes are preserved, so the walk plays exactly as composed — just somewhere else.
 */
class WalkTransposition(anchor: WalkAnchor, listener: LatLngD, heading: Double) {
    private val from: LatLngD = anchor.coord
    private val to: LatLngD = listener
    /** Radians clockwise: listener heading minus anchor heading. */
    private val turn: Double = (heading - anchor.heading) * Math.PI / 180

    companion object {
        private const val METRES_PER_DEGREE_LAT = 110_540.0
        private fun metresPerDegreeLng(lat: Double) = 111_320.0 * cos(lat * Math.PI / 180)
    }

    fun apply(c: LatLngD): LatLngD {
        // Local east/north offset from the anchor, in metres.
        val east = (c.lng - from.lng) * metresPerDegreeLng(from.lat)
        val north = (c.lat - from.lat) * METRES_PER_DEGREE_LAT
        // Turn the bearing of that offset by `turn` (clockwise from north).
        val cs = cos(turn); val sn = sin(turn)
        val east2 = east * cs + north * sn
        val north2 = north * cs - east * sn
        return LatLngD(
            lat = to.lat + north2 / METRES_PER_DEGREE_LAT,
            lng = to.lng + east2 / metresPerDegreeLng(to.lat),
        )
    }

    fun apply(pair: List<Double>): List<Double> {
        if (pair.size != 2) return pair
        val c = apply(LatLngD(pair[0], pair[1]))
        return listOf(c.lat, c.lng)
    }
}

/**
 * A copy of this walk moved and turned onto the listener. Radii are untouched — only positions
 * move — so every area keeps its authored size and relative bearing.
 *
 * Labels are deliberately left where they were authored, matching iOS. See the note in the Android
 * port README: this looks like an iOS bug, but parity wins until the iOS side changes.
 */
fun SoundMap.transposed(t: WalkTransposition): SoundMap = copy(
    center = center?.let { t.apply(it) },
    shapes = shapes.map { s ->
        s.copy(
            center = s.center?.let { t.apply(it) },
            points = s.points?.map { t.apply(it) },
        )
    },
    // Routes travel with the walk too — a suggested path left behind where the walk was authored
    // would point the listener at nothing.
    routes = routes?.map { r -> r.copy(points = r.points.map { t.apply(it) }) },
)

/**
 * A lat/lng shift used by "re-center map over me", so a walk authored for one city can be tested
 * wherever the user physically is.
 */
data class CoordinateOffset(val dLat: Double = 0.0, val dLng: Double = 0.0) {
    fun apply(c: LatLngD): LatLngD = LatLngD(c.lat + dLat, c.lng + dLng)

    val isZero: Boolean get() = abs(dLat) < 1e-12 && abs(dLng) < 1e-12

    companion object {
        val NONE = CoordinateOffset()

        /** Offset that moves [mapCenter] on top of [userLocation]. */
        fun recentering(mapCenter: LatLngD, onto: LatLngD) =
            CoordinateOffset(onto.lat - mapCenter.lat, onto.lng - mapCenter.lng)
    }
}
